### ISSUE TITLE
Offset issues

### ISSUE THREAD

**Issue by hellocatfood** [fablabnbg/inkscape-silhouette#15]
I'm trying to send the attached file to a Sillhouette Cameo machine. It's A3 size and has nothing else in it. When it's sent, however, there is a very odd offset that results in the drawing coming off the page (image attached). I'm not using a cutting may (going straight to card) and have adjusted my drawing to account for the area lost when not using one.

[attachments.zip](https://github.com/fablabnbg/inkscape-silhouette/files/71070/attachments.zip)

**Comment by tuxyme** [comment]
You are using a Cameo2, with device id 0b4d:112b ?

**Comment by hellocatfood** [comment]
I'm using the extension again and recently noticed that all of my cuts are off by 10mm. I have no offset set in my settings (screenshot attached)

![screenshot from 2017-04-25 00-25-10](https://cloud.githubusercontent.com/assets/151947/25362735/b786d000-294d-11e7-9ff2-cdc0d4894649.png)

**Comment by ChristianZi** [comment]
@hellocatfood have you figured out the reason for the 10mm offset? 
Everything I do is shiftet about 10mm to the right.
My Silhouette Cameo has the device-Id 0b4d:1121

**Comment by hellocatfood** [comment]
@ChristianZi sadly I didn't solve it and the problem still exists

**Comment by jnweiger** [comment]
Does a manual X offset: -10mm bring it back to normal?

**Comment by ChristianZi** [comment]
@jnweiger: what I have tested is, setting an Offset in the UI. 
I was not happy with the result, but I do not remember what the result was.
I hope I will find some time the next days to test it again.

**Comment by hellocatfood** [comment]
@jnweiger in the past when I've tried that 10mm gets cut off from the right of the artwork

**Comment by ChristianZi** [comment]
@jnweiger when I do an offset of -10, I'm not able to do anything in the left 10 mm.

**Comment by EtherGraf** [comment]
An offset to which origin? Couldn't you simply place the paper 10mm to the right?
Maybe your cutter device has some pointers missadjusted?
Does it happen with A3 only or A4 landscape or portrait?

**Comment by hellocatfood** [comment]
@jnweiger @EtherGraf Still happening as of August 2018. I've tried this with A5, A4, A3 at both landscape and portrait.

**Comment by hellocatfood** [comment]
I wanted to provide further illustration that can help identify the problem.

I'm using the Silhouette Cameo 2 with device id 0b4d:112b. 

I am using both the official cutting mat from Silhouette and one from Circut. Both mats are the same size. As you can see in the photo below the Circut mat only aligns horizontally with the Silhouette mat at the shaded white area, not where the grid is. The same happens vertically.

![2018-08-04 11 52 21](https://user-images.githubusercontent.com/151947/43687905-39105576-98d6-11e8-97fc-5d03a40eed2b.jpg)

If I place a paper (in this photo A5 in portrait) on the mat in the top left corner then the artwork goes approximately 10mm over on the right.

![2018-08-04 11 53 48](https://user-images.githubusercontent.com/151947/43687913-5cb8b284-98d6-11e8-9861-57220fb3a36d.jpg)

If I move the paper itself approximately 5mm to the right then the artwork draws in the right place. In the below photo the blue marks are from when it was placed at the edge of the grid and the orange when it was place 5mm to the right.

![2018-08-03 11 32 19](https://user-images.githubusercontent.com/151947/43687920-81112468-98d6-11e8-860a-1ca9d3d847d7.jpg)

Although this "works" placing the paper 10mm to the right is imprecise and it means that 10mm is lost on the overall drawing. Not a lot but still a waste.

As said before using an offset in the extensions options of -10mm on the X axis results in the artwork getting cut off.

I also tried moving the cutting mat itself 10mm to the right but that means it misses the rollers (on the Cameo 2 there's different position markers for the rollers but only on the right side).

![2018-08-04 12 24 25](https://user-images.githubusercontent.com/151947/43687923-a06fd912-98d6-11e8-9a9d-1fe9e6ba1e4a.jpg)

I found a Windows machine and tried the same drawing using Silhouette's software and didn't experience any issues.

**Comment by mountaineermaniac** [comment]
Having similar issues here.. Linux Mint19/Inkscape 0.92/Cemeo2, however my x and Y "0" are actually about 14mm high of the grid on the mat and 73mm to the left.  I have tried using a changing the offsets to correct this, but when i put in a negative offset, cameo goes thru some motion and print/cuts nothing.  Not sure how I can move the "0" to align with true position of cutter.

**Comment by EtherGraf** [comment]
Hello @hellocatfood,

thanks for your detailed investigation. At the moment I cannot spend any time to dig into this. But a quick glance to the code shows me:

 'margin_left_mm':9.0, 'margin_top_mm':1.0

at Graphtec.py, line 137 for the cameo2 model. I have no idea of the background of this setting, but the following code does some calculating of the ploting area based on this value.

Could you temporarily change this value to 0.0 and give it a try?

@jnweiger: Do you remember the meaning of this device setting and can give us an explanation?

Setting an additional positive offset in the settings does currently cut of the artwork on the right side. This is by (faulty?) design, because the artwork gets shifted on the given plot area. The plot area is limited by device and document size. A workaround for this is to increase the document size.

**Comment by hellocatfood** [comment]
I changed the code and it is a lot closer to working. Below I set the page to 148mmx210mm and put a circle in a square (convert to paths) in two corners. I used an A4 paper so that I could see the results better.

![img_20180909_185320](https://user-images.githubusercontent.com/151947/45270044-052e2280-b490-11e8-9d44-79c51586d41e.jpg)

As you can the bottom-left square isn't plotted and the circle has bits missing on the bottom and left (I think the missing bit on the top-right is the pen being weird). I tried again on paper sized 148mmx210mm and got the same result (the pink line is the one plotted with the modified code).

![img_20180910_002324](https://user-images.githubusercontent.com/151947/45270047-0cedc700-b490-11e8-9638-c5a4eaa86d14.jpg)

**Comment by jnweiger** [comment]
Back in the days when there was only cameo 1, I had margin trim and offset and all that working correctly. I fear something was added later that now spoils the fun.

**Comment by hellocatfood** [comment]
@jnweiger I'm happy to help with debugging. Is there anything else you can suggest I change in the code? I'll try some other things on line 137.

**Comment by EtherGraf** [comment]
Seems user @dbajar is on the right track with commits fe9a82df67bb3a8cf09dfa7d8ef1f6aa97dc3fdb and 7d96cfda65074c50db8752af63d84b4c8126d43e

**Comment by dbajar** [comment]
Yes, I had similar offset and cropping issues as reported in this thread and others.  I've added some changes to https://github.com/dbajar/inkscape-silhouette that appear to fix things on Cameo3.

Unfortunately I don't have any other devices to test with.

**Comment by miLORD1337** [comment]
> Back in the days when there was only cameo 1, I had margin trim and offset and all that working correctly. I fear something was added later that now spoils the fun.

Actually, on my Cameo 1 now the margins are off by almost exactly the hard-coded values. But when I insert negative values into the dialog box in Inkscape, the plotter skips the entire cut. Therefore I updated the values to `'margin_left_mm':-0.5, 'margin_top_mm':-0.25` to adjust it to my preferences.

Speaking of preferences: Maybe a "Preferences" page would be useful to make those values editable instead of keeping them hard coded? Theoretically it wouldn't matter too much, as we have the offset input as well, but with only non-negative numbers those cannot be used for all cases.

_Or is there an option to recalibrate my plotter, so the margins are correct again?_

**Comment by tiger12506** [comment]
I recently have experienced this issue (inkscape-silhouette ver 1.29) on a Silhouette Cameo 1. The artificially imposed margins are about 5mm or so on the left, right, and bottom sides (NOT the top side). I have an X-offset = -5mm, and Y-offset = 5mm, to get the origin point in the correct place on the cutting mat I'm currently using.

I haven't tested changing the hardcoded offset values, but I will be sure to report back on the situation in a few days time.
This is currently a show-stopper for me because I am cutting a large vinyl print (about 12 sq ft) and piecing it together. The false margins are a huge problem as I previously split my image into 12" squares exactly.

**Comment by miLORD1337** [comment]
Just to be sure @tiger12506, you are saying that you can set negative offsets and it still makes a cut there? This is not the case for me. It only works when adjusting the hardcoded values. 

Any news regarding that already?

Edit: I created a pull request now, because I ran into the same problem again after updating the extension. See #316

**Comment by tiger12506** [comment]
> Just to be sure @tiger12506, you are saying that you can set negative offsets and it still makes a cut there? This is not the case for me. It only works when adjusting the hardcoded values.

No, setting negative values does not make a cut near the margin, but it fixes the alignment of the design as if it would have made the margin cuts. I haven't tested it with modified hardcoded values yet, life got in the way. Thanks for looking into it, I'll test the pull request when I next get an opportunity.
